### PR TITLE
Allow ingesting Avro complex types into JSON column.

### PR DESCRIPTION
**Pull request by amrishlal**
## Description
This PR adds support for type conversion from avro complex types (MAP, ARRAY, RECORD, UNION, FIXED, ENUM) to Pinot JSON column. Currently, only avro type STRING conversion to JSON column is supported. Ingestion of complex types into JSON column can also be done using Ingestion Transform Function; however, this creates issues for us as the Pinot column name can't be of the same name as the input avro field name.

The main changes, that allows for ingesting complex types into JSON, is in `DataTypeTransformer.java` and `AvroSchemaUtil.java`. `PinotDataType.java` contains a bug fix (the fact that a simple string is also a valid json string) which allows for proper conversion of STRING field to JSON column.

`AvroIngestionSchemaValidator` was modified to make sure that we no longer reject avro complex type (MAP, ARRAY, RECORD, UNION) field if they are being ingested into JSON column.

The rest of the files (`PinotDataTypeTest.java`, `JsonIngestionFromAvroQueriesTest.java`, `AvroRecordExtractorTest.java`, and `AbstractRecordExtractorTest.java`) contain test cases to validate ingestion of avro complex types into JSON column.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import org.apache.pinot.spi.data.readers.RecordReader;
+import org.apache.pinot.spi.utils.Pair;
+import org.apache.pinot.spi.utils.ReadMode;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.apache.avro.Schema.*;
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:61]: Seems there are more usage of avro Schema than pinot Schema. Consider import avro Schema and use full path for pinot Schema

File: pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java [hunk 2]
```diff
@@ -829,18 +829,23 @@ public String toString(Object value) {
   public String toJson(Object value) {
     if (value instanceof String) {
       try {
+        // Try to parse the string as JSON first
         return JsonUtils.stringToJsonNode((String) value).toString();
+      } catch (com.fasterxml.jackson.core.JsonParseException jpe) {
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/PinotDataType.java:834]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+  private static Map<String, String> createMapField(Pair<String, String>[] pairs) {
+    Map<String, String> map = new HashMap<>();
+    for (Pair<String, String> pair : pairs) {
+      map.put(pair.getFirst(), pair.getSecond());
+    }
+    return map;
+  }
+
+  private static org.apache.avro.Schema createRecordSchema() {
+    List<org.apache.avro.Schema.Field> fields = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:121]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java (hunk 1 above)
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:61]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+      Assert.assertEquals(block.getDataSchema().getColumnDataType(1), DataSchema.ColumnDataType.STRING);
+
+      List<String> expecteds =
+          Arrays.asList("[1, null]", "[2, null]", "[3, null]", "[4, null]", "[5, minney]", "[6, null]", "[7, scooby]");
+      int index = 0;
+
+      Iterator<Object[]> iterator = rows.iterator();
+      while (iterator.hasNext()) {
+        Object[] row = iterator.next();
+        System.out.println(Arrays.toString(row));
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:267]: Done.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        Assert.assertEquals(Arrays.toString(row), expecteds.get(index++));
+      }
+    } catch (IllegalStateException ise) {
+      Assert.assertTrue(true);
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:271]: Fixed.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroIngestionSchemaValidator.java [hunk 6]
```diff
@@ -124,13 +124,16 @@ private void validateSchemas() {
...
         }
-        FieldSpec.DataType dataTypeForSVColumn = AvroUtils.extractFieldDataType(avroColumnField);
-        // check data type mismatch
-        if (fieldSpec.getDataType() != dataTypeForSVColumn) {
-          _dataTypeMismatch.addMismatchReason(String
-              .format("The Pinot column: (%s: %s) doesn't match with the column (%s: %s) in input %s schema.",
-                  columnName, fieldSpec.getDataType().name(), avroColumnName, avroColumnType.name(),
-                  getInputSchemaType()));
+
+        if (fieldSpec.getDataType() != FieldSpec.DataType.JSON) {
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroIngestionSchemaValidator.java:128]: Fixed.

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.apache.avro.Schema.*;
+
+
+/**
+ * Test if ComplexType (RECORD, ARRAY, MAP, and UNION) field from an AVRO file can be ingested into a JSON column in
+ * a Pinot segment. This class tests. Ingestion from ENUM (symbol) and FIXED (binary) is not supported.
```
[the remark this answers is not in the document]
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java]: Added tests for verifying support for ingesting FIXED and ENUM into JSON column.

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 8]
```diff
@@ -95,6 +100,9 @@ public static ObjectNode toAvroSchemaJsonObject(FieldSpec fieldSpec) {
       case BYTES:
         jsonSchema.set("type", convertStringsToJsonArray("null", "bytes"));
         return jsonSchema;
+      case JSON:
+        jsonSchema.set("type", convertStringsToJsonArray("null", "json"));
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:104]: Should this be `string`? I don't think avro has `json` type

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java [hunk 9]
```diff
@@ -50,6 +50,11 @@ public static DataType valueOf(Schema.Type avroType) {
         return DataType.STRING;
       case BYTES:
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:51]: FIXED can be handled as BYTES

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+
+  private static GenericData.Fixed createFixedField(Schema fixedSchema, int value) {
+    byte[] bytes = {(byte) (value >> 24), (byte) (value >> 16), (byte) (value >> 8), (byte) value};
+    return new GenericData.Fixed(fixedSchema, bytes);
+  }
+
+  private static void createInputFile()
+      throws IOException {
+    INDEX_DIR.mkdir();
+    Schema avroSchema = Schema.createRecord("eventsRecord", null, null, false);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:152]: (minor) Either remove `Schema.` or remove the static import

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorTest.java [hunk 11]
```diff
@@ -73,8 +73,12 @@ protected void createInputFile()
...
             new Field("lastName", createUnion(Lists.newArrayList(create(Type.STRING), create(Type.NULL))), null, null),
             new Field("bids", createUnion(Lists.newArrayList(createArray(create(Type.INT)), create(Type.NULL))), null,
-                null), new Field("campaignInfo", create(Type.STRING), null, null),
-            new Field("cost", create(Type.DOUBLE), null, null), new Field("timestamp", create(Type.LONG), null, null));
+                null),
+            new Field("campaignInfo", create(Type.STRING), null, null),
+            new Field("cost", create(Type.DOUBLE), null, null),
+            new Field("timestamp", create(Type.LONG), null, null),
+            new Field("xarray", createArray(create(Type.STRING))),
+            new Field("xmap", createMap(create(Type.STRING))));
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorTest.java:81]: Suggest adding all supported types (BOOLEAN, ENUM, FIXED, RECORD)

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java (hunk 8 above)
Comment (amrishlal) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:104]: Done

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java (hunk 9 above)
Comment (amrishlal) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/main/java/org/apache/pinot/plugin/inputformat/avro/AvroSchemaUtil.java:51]: Done

File: pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java (hunk 10 above)
Comment (amrishlal) [reply pinot-core/src/test/java/org/apache/pinot/queries/JsonIngestionFromAvroQueriesTest.java:152]: Done

File: pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorTest.java (hunk 11 above)
Comment (amrishlal) [reply pinot-plugins/pinot-input-format/pinot-avro-base/src/test/java/org/apache/pinot/plugin/inputformat/avro/AvroRecordExtractorTest.java:81]: This file is shared by Thrift, Protobuf, ORC, CSV, etc. and I am seeing some differences in handling of boolean (Thrift vs others), fixed (custom code needed for setting Fixed values in Avro, vs binary in others), and enum (custom code needed for Avro) between different file formats, so probably a better idea to do this in a separate PR of its own.

**Comment by amrishlal** [comment]
@siddharthteotia please review.

**Pull request merged**